weighted_graphs.py

```diff
@@ -0,0 +1,214 @@
+      # ^ not sure what that means and why it's important we do this
+      # not terribly sure why curr_cost_to_start_vert might be more than
+      # costs_to_start_vert[curr_vert]
+      if curr_cost_to_start_vert > costs_to_start_vert[curr_vert]:
+        print(f"curr_cost_to_start_vert: {curr_cost_to_start_vert} > costs_to_start_vert[curr_vert]: {costs_to_start_vert[curr_vert]}, curr_vert: {curr_vert}")
+        continue
+
+      vertexes_directly_connected_to_curr_vert = [vertex for vertex, _cost in graph[curr_vert].items()]
+      print(f"{curr_vert} is connected to {vertexes_directly_connected_to_curr_vert}")
+      # neighbour vertex means a vertex that is directly connected to our current vertex
+      for neighbour_vertex, cost in graph[curr_vert].items():
+        # let's say it cost `7` to get to vertex `c`
+        # `c` is directly connected to `d` and the cost of c -> d is 3
+        # so proposed_cost to go from start_vert to `d` == 7 +3 == 10
+        proposed_cost = curr_cost_to_start_vert + cost
+
+        if proposed_cost < costs_to_start_vert[neighbour_vertex]:
+        # found a shorter path to this neighbour_vertex! update the dict!
+          print(f"from {starting_vertex} to {neighbour_vertex}, proposed cost {proposed_cost} < current cost {costs_to_start_vert[neighbour_vertex]}. update dict")
+          costs_to_start_vert[neighbour_vertex] = proposed_cost
+
+          # add this to the priority queue
+          heapq.heappush(priority_queue, (proposed_cost, neighbour_vertex))
+
+        else:
+          print(f"from {starting_vertex} to {neighbour_vertex}, proposed cost {proposed_cost} >= current cost {costs_to_start_vert[neighbour_vertex]}. don't update dict")
+
+        print(f"processed neighbour vertex {neighbour_vertex}. costs_to_{starting_vertex}: {costs_to_start_vert})\n")
+
+    print(f"\n**ending**\nfinal costs to each vertex from vertex {starting_vertex}: {costs_to_start_vert}")
+    print(f"original graph: {graph}")
+    return costs_to_start_vert
+
+  def network_delay_time(self, times, N, K):
+    """
+      from https://leetcode.com/problems/network-delay-time/
+      There are N network nodes, labelled 1 to N.
+
+      Given times, a list of travel times as directed edges
+      times[i] = (u, v, w), where u is the source node, v is the target node,
+      and w is the time it takes for a signal to travel from source to target.
+
+      Now, we send a signal from a certain node K.
+
+      How long will it take for all nodes to receive the signal?
+      If it is impossible, return -1.
+
+      times: List[List[int]], N: int, K: int) -> int:
+
+      Example
+      Input: times = [[2,1,1],[2,3,1],[3,4,1]], N = 4, K = 2
+      Output: 2
+
+      Runtime: 496 ms, faster than 73.14% of Python3 online submissions
+      for Network Delay Time.
+      Memory Usage: 15.8 MB, less than 19.89% of Python3 online submissions
+      for Network Delay Time.
```

Comment:
there's probably a better way to do this since i'm only in the top 27%? can't think of any way right now. probably involves simplifying the first step of building the graph or finding the max time cos i can't think of any other way to simplify dijkstra's and we need dijkstra's because we're trying to find a path from one node to every other node so that we can find the max time.